db/database.go

```diff
@@ -878,40 +880,50 @@ func (db *Database) Compact() (int, error) {
 	purgedDocs := make([]string, 0)
 
-	var tombstonesRow QueryIdRow
-	for results.Next(&tombstonesRow) {
-		base.InfofCtx(ctx, base.KeyCRUD, "\tDeleting %q", tombstonesRow.Id)
-		// First, attempt to purge.
-		purgeErr := db.Purge(tombstonesRow.Id)
-		if purgeErr == nil {
-			purgedDocs = append(purgedDocs, tombstonesRow.Id)
-		} else if base.IsKeyNotFoundError(db.Bucket, purgeErr) {
-			// If key no longer exists, need to add and remove to trigger removal from view
-			_, addErr := db.Bucket.Add(tombstonesRow.Id, 0, purgeBody)
-			if addErr != nil {
-				base.WarnfCtx(ctx, base.KeyAll, "Error compacting key %s (add) - tombstone will not be compacted.  %v", base.UD(tombstonesRow.Id), addErr)
-				continue
-			}
+	ctx := db.Ctx
+
+	for {
+		results, err := db.QueryTombstones(purgeOlderThan, QueryTombstoneBatch, gocb.RequestPlus)
+		if err != nil {
+			return 0, err
+		}
 
-			// At this point, the doc is not in a usable state for mobile
-			// so mark it to be removed from cache, even if the subsequent delete fails
-			purgedDocs = append(purgedDocs, tombstonesRow.Id)
+		base.InfofCtx(ctx, base.KeyAll, "Compacting purged tombstones for %s ...", base.MD(db.Name))
+		purgeBody := Body{"_purged": true}
+		var tombstonesRow QueryIdRow
+		var resultCount int
+		for results.Next(&tombstonesRow) {
+			resultCount++
+			base.InfofCtx(ctx, base.KeyCRUD, "\tDeleting %q", tombstonesRow.Id)
+			// First, attempt to purge.
+			purgeErr := db.Purge(tombstonesRow.Id)
+			if purgeErr == nil {
+				purgedDocs = append(purgedDocs, tombstonesRow.Id)
+			} else if base.IsKeyNotFoundError(db.Bucket, purgeErr) {
+				// If key no longer exists, need to add and remove to trigger removal from view
+				_, addErr := db.Bucket.Add(tombstonesRow.Id, 0, purgeBody)
+				if addErr != nil {
+					base.WarnfCtx(ctx, base.KeyAll, "Error compacting key %s (add) - tombstone will not be compacted.  %v", base.UD(tombstonesRow.Id), addErr)
+					continue
+				}
 
-			if delErr := db.Bucket.Delete(tombstonesRow.Id); delErr != nil {
-				base.ErrorfCtx(ctx, base.KeyAll, "Error compacting key %s (delete) - tombstone will not be compacted.  %v", base.UD(tombstonesRow.Id), delErr)
+				// At this point, the doc is not in a usable state for mobile
+				// so mark it to be removed from cache, even if the subsequent delete fails
+				purgedDocs = append(purgedDocs, tombstonesRow.Id)
+
+				if delErr := db.Bucket.Delete(tombstonesRow.Id); delErr != nil {
+					base.ErrorfCtx(ctx, base.KeyAll, "Error compacting key %s (delete) - tombstone will not be compacted.  %v", base.UD(tombstonesRow.Id), delErr)
+				}
+			} else {
+				base.WarnfCtx(ctx, base.KeyAll, "Error compacting key %s (purge) - tombstone will not be compacted.  %v", base.UD(tombstonesRow.Id), purgeErr)
 			}
-		} else {
-			base.WarnfCtx(ctx, base.KeyAll, "Error compacting key %s (purge) - tombstone will not be compacted.  %v", base.UD(tombstonesRow.Id), purgeErr)
+		}
+
+		if resultCount < QueryTombstoneBatch {
+			break
```

Comment:
As discussed, move the 'purge them from all channel caches' inside the loop, to avoid purgedDocs chewing up a lot of memory.